update ada-url to v3.2.2

anonrig opened the pull request:
I'm one of the co-authors of Ada URL parser. I saw that Redpanda still uses an extremely old version of Ada. I recommend upgrading it.

anonrig committed:
update ada-url to v3.2.2

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=25790) <br/>All committers have signed the CLA.

rockwotj commented:
Thanks @anonrig - actually our public CMake build is deprecated, we're switching to Bazel. So that is why this is so old

rockwotj commented:
although we're using an old version there too so will upgrade.

anonrig commented:
> although we're using an old version there too so will upgrade.

Should I update this PR, or will you?

anonrig commented:
> > although we're using an old version there too so will upgrade.
> 
> Should I update this PR, or will you?

Actually, it seems you need to update `vectorized-public` S3 bucket with 3.2.2 in order to update Bazel. You can also use "singleheader.zip" file from Github releases with fixed shasum (just like Cloudflare) - https://github.com/cloudflare/workerd/blob/main/build/deps/gen/dep_ada_url.bzl#L6

rockwotj commented:
Yes exactly, I will do it Monday. Thanks again! Switching to single header is a great idea

anonrig commented:
> Yes exactly, I will do it Monday. Thanks again! Switching to single header is a great idea

@rockwotj I've updated the Bazel file to help you. Feel free to close this or use as a baseline.

rockwotj commented:
> > Yes exactly, I will do it Monday. Thanks again! Switching to single header is a great idea
> 
> 
> 
> @rockwotj I've updated the Bazel file to help you. Feel free to close this or use as a baseline.

Fantastic, I will just push the updated URL to this PR when I can mirror the zip. This way you are properly attributed.

anonrig commented on MODULE.bazel.lock:
```diff
@@ -5774,12 +5774,12 @@
         "usagesDigest": "iKmmEQX8xqicrHiRHLqXwjwu7pCrLkIAFXGNdc4Aveo=",
         "recordedFileInputs": {
-          "@@rules_rust~~rust_host_tools~rust_host_tools//bin/rustc": "241027b94ad67beb36d0356f7cc05180daaa9966b2958cfba1d4de089d2e521c",
+          "@@rules_rust~~rust_host_tools~rust_host_tools//bin/rustc": "b953c69e6f682bf55e39bdd55bdcbd3e1cc0cbdf9dfda24b9c6478ee2c38359b",
```
Note: I've no idea why this line has changed. Please review accordingly.

rockwotj replied:
It's a known issue, I will fixup

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--6fdcb5c6-5698-479a-88ae-5ab745562bc1-->
<!--TEST RESULTS-->
<!--426408fc-1b13-4a35-9382-b3e7d2b02007-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64456">build#64456</a></summary>


| test_id                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit                                                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/64456#01962ba1-ae4d-458f-acae-629b2d26e8fb | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64456#01962bfc-29a3-47b9-9722-7e2599eee8b8 | FLAKY         | 1/2      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65643">build#65643</a></summary>


| test_id                                                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.internal_topic_protection_test.InternalTopicProtectionLargeClusterTest.test_consumer_offset_topic | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65643#0196abd9-5bf4-4598-b896-3780bc423349 | FLAKY         | 20/21    |
</details>
<!--/body-->

rockwotj committed:
bazel: upgrade ada to latest

Also switch to the "singleheader" version of ada, which is trivial to
build in Bazel instead of shelling out to cmake.

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/11d2fa5b-fac0-404b-ba6b-6f32eae1a4e7?refs=d45926b61a3a45dd2e6cea2cc0079c2b334b8db9&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/11d2fa5b-fac0-404b-ba6b-6f32eae1a4e7/license?refs=d45926b61a3a45dd2e6cea2cc0079c2b334b8db9&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/a7754b06-52ed-4497-ab4f-5032d2d64641?refs=d45926b61a3a45dd2e6cea2cc0079c2b334b8db9&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/a7754b06-52ed-4497-ab4f-5032d2d64641/license?refs=d45926b61a3a45dd2e6cea2cc0079c2b334b8db9&source=prChecksComment)

anonrig commented:
@rockwotj we fixed a bug with a flag in Ada which might cause issues in this pull request as well. Since I can't see the build logs, can you try https://github.com/ada-url/ada/releases/tag/v3.2.4 and see if the errors are gone?

anonrig commented on src/v/thirdparty/ada/BUILD:
```diff
@@ -1,5 +1,6 @@
 cc_library(
     name = "ada",
+    srcs = [],
```
Ada will not work without the cpp file.

rockwotj replied:
This doesn't quite do what you're thinking. This is just a way to forward the header - I don't think we quite need this anymore that we're using a native bazel build, but will fix that later.

anonrig commented on bazel/thirdparty/ada.BUILD:
```diff
@@ -1,27 +1,11 @@
-    cache_entries = {
-        "BUILD_SHARED_LIBS": "OFF",
-        "ADA_TESTING": "OFF",
-        "ADA_TOOLS": "OFF",
```
We need Ada include url pattern 0 define statement to reduce the binary size.

rockwotj replied:
Thanks, we don't really care too much about binary size, but I've removed it anyways

rockwotj commented:
The build issues are due to our legacy build system which is not defined in this repo and since there are breaking changes in the v3 release, I will have to wait until we remove it (we're working on it) until I can merge this PR. I am not sure how long it will take to remove, but I have not forgotten about this PR

anonrig commented:
Ah yes. V3 uses C++20. Thanks for the update!

rockwotj commented:
/ci-repeat

rockwotj committed:
bazel/ada: disable features we don't use

rockwotj committed:
thirdparty: upgrade to latest ada version (3.2.4)

anonrig commented on cmake/dependencies.cmake:
```diff
@@ -95,9 +95,10 @@ fetch_dep(GTest
 fetch_dep(ada
   REPO https://github.com/ada-url/ada
-  TAG v2.7.3)
+  TAG v3.2.2)
```
This should be 3.2.4 I suppose?

anonrig replied:
Cc @rockwotj

rockwotj replied:
This file is unused and vestigial. Will remove in a followup PR

dotnwat commented:
nice. thanks @anonrig
